### PR TITLE
Deprecate single component extraction methods in libcudf

### PR DESCRIPTION
**Pull request by Matt711**
## Description

This PR deprecates the single component extraction methods (eg. `cudf::datetime::extract_year`) that are already covered by `cudf::datetime::extract_datetime_component`.
xref #17143 
## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by wence-** [comment]
Marked as breaking because this is a change that requires downstream users to do something.

**Comment by Matt711** [comment]
> Marked as breaking because this is a change that requires downstream users to do something.

What do they have to do?

File: cpp/tests/datetime/datetime_ops_test.cpp [hunk 1]
```diff
@@ -52,16 +52,6 @@ TYPED_TEST(NonTimestampTest, TestThrowsOnNonTimestamp)
   cudf::data_type dtype{cudf::type_to_id<T>()};
   cudf::column col{dtype, 0, rmm::device_buffer{}, rmm::device_buffer{}, 0};
 
-  EXPECT_THROW(extract_year(col), cudf::logic_error);
```
Comment (bdice) [inline cpp/tests/datetime/datetime_ops_test.cpp:55]: I don't think we should delete tests until the corresponding code is removed.

Comment (davidwendt) [reply cpp/tests/datetime/datetime_ops_test.cpp:55]: We should delete the tests but in this case they should be replaced with the equivalent calls to the new API in this PR.

File: cpp/tests/datetime/datetime_ops_test.cpp [hunk 2]
```diff
@@ -104,96 +94,6 @@ TEST_F(BasicDatetimeOpsTest, TestExtractingDatetimeComponents)
       987234623   // 1970-01-01 00:00:00.987234623 GMT
     };
 
-  CUDF_TEST_EXPECT_COLUMNS_EQUAL(*extract_year(timestamps_D),
```
Comment (davidwendt) [inline cpp/tests/datetime/datetime_ops_test.cpp:107]: These should be changed to call the `extract_datetime_component` and not just removed.

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/tests/datetime/datetime_ops_test.cpp (hunk 2 above)
Comment (wence-) [reply cpp/tests/datetime/datetime_ops_test.cpp:107]: I think this is already happening below: the new usage is already tested: https://github.com/rapidsai/cudf/blob/7ea8f32877ec59197870d39a54c40bd9d744dade/cpp/tests/datetime/datetime_ops_test.cpp#L97

File: cpp/tests/datetime/datetime_ops_test.cpp [hunk 3]
```diff
@@ -333,123 +233,6 @@ struct TypedDatetimeOpsTest : public cudf::test::BaseFixture {
 
 TYPED_TEST_SUITE(TypedDatetimeOpsTest, cudf::test::TimestampTypes);
 
-TYPED_TEST(TypedDatetimeOpsTest, TestEmptyColumns)
```
Comment (wence-) [inline cpp/tests/datetime/datetime_ops_test.cpp:336]: I think these ones want to use the new `extract_datetime_component` mechanism

File: cpp/tests/datetime/datetime_ops_test.cpp (hunk 2 above)
Comment (Matt711) [reply cpp/tests/datetime/datetime_ops_test.cpp:107]: Yep, this test uses `extract_datetime_component` already. But I added the others ones I deleted back

File: cpp/tests/datetime/datetime_ops_test.cpp (hunk 1 above)
Comment (Matt711) [reply cpp/tests/datetime/datetime_ops_test.cpp:55]: Done.

**Comment by wence-** [comment]
> > Marked as breaking because this is a change that requires downstream users to do something.
> 
> What do they have to do?

Migrate to the new interface (to avoid compile warnings)

File: cpp/tests/datetime/datetime_ops_test.cpp (hunk 1 above)
Comment (bdice) [reply cpp/tests/datetime/datetime_ops_test.cpp:55]: This is marked as resolved but I don't see any equivalent test in this file. We want a test that ensures the new date extraction function throws a `cudf::logic_error` if given a non-timestamp input. If I'm just not seeing this, can you link to it in the source?

Comment (Matt711) [reply cpp/tests/datetime/datetime_ops_test.cpp:55]: Ah you're right.

Comment (Matt711) [reply cpp/tests/datetime/datetime_ops_test.cpp:55]: Added those tests back.

**Review by davidwendt — APPROVED** [review]

**Review by karthikeyann — APPROVED** [review]

**Pull request merged**